<a href="https://colab.research.google.com/github/iharshidas/Widhya_winternship/blob/main/Bitcoin_price_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Bitcoin Price Prediction (Data Science)**
### **By: Harshita Das**
Widhya Winternship Mission-1

In [39]:
# Importing the required libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

In [40]:
# Reading the data from a url link into a dataframe
link='https://raw.githubusercontent.com/WidhyaOrg/datasets/master/bitcoin_dataset.csv'
bpp_df=pd.read_csv(link)
print('Number of rows and columns: ',bpp_df.shape)
bpp_df.head()

Number of rows and columns:  (2906, 24)


Date  ...  btc_estimated_transaction_volume_usd
0  2/17/2010  ...                                   0.0
1  2/18/2010  ...                                   0.0
2  2/19/2010  ...                                   0.0
3  2/20/2010  ...                                   0.0
4  2/21/2010  ...                                   0.0

[5 rows x 24 columns]

In [41]:
# Printing the 1024th value in 'btc_market_price' column
print(bpp_df['btc_market_price'].iloc[1023])

13.6888


In [42]:
# Finding the column which shows highest correlation with the 'btc_market_price' column
bpp_df[bpp_df.columns[1:]].corr()['btc_market_price'][:]

btc_market_price                                       1.000000
btc_total_bitcoins                                     0.392474
btc_market_cap                                         0.999783
btc_trade_volume                                       0.868813
btc_blocks_size                                        0.674507
btc_avg_block_size                                     0.544033
btc_n_orphaned_blocks                                 -0.076400
btc_n_transactions_per_block                           0.544125
btc_median_confirmation_time                           0.275250
btc_hash_rate                                          0.920876
btc_difficulty                                         0.916937
btc_miners_revenue                                     0.986732
btc_transaction_fees                                   0.806595
btc_cost_per_transaction_percent                      -0.013517
btc_cost_per_transaction                               0.815507
btc_n_unique_addresses                  

In the above correlation, we find that **"btc_market_cap" has the highest correlation to "btc_market_price" i.e., of 0.999783**

In [43]:
bpp_new=bpp_df[['btc_market_price','btc_market_cap','btc_n_transactions','btc_n_transactions','btc_cost_per_transaction','btc_difficulty','btc_hash_rate','btc_cost_per_transaction_percent']].copy()
bpp_new.head()

btc_market_price  ...  btc_cost_per_transaction_percent
0               0.0  ...                         31.781022
1               0.0  ...                        154.463801
2               0.0  ...                       1278.516635
3               0.0  ...                      22186.687990
4               0.0  ...                        689.179876

[5 rows x 8 columns]

In [44]:
# Importing the modelling libraries
from sklearn.linear_model import LinearRegression  
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.impute import SimpleImputer as Imp

In [56]:
# Dividing data into target and features
y=bpp_new.iloc[:,0].values.reshape(-1, 1)
X=bpp_new.iloc[:,1:].values

In [46]:
bpp_new.isnull().sum()

btc_market_price                     0
btc_market_cap                       0
btc_n_transactions                   0
btc_n_transactions                   0
btc_cost_per_transaction             0
btc_difficulty                      16
btc_hash_rate                        0
btc_cost_per_transaction_percent     0
dtype: int64

In [47]:
imp=Imp(missing_values=np.nan, strategy='mean')
imp.fit(bpp_new[['btc_difficulty']])
bpp_new['btc_difficulty']=imp.transform(bpp_new[['btc_difficulty']])

In [48]:
bpp_new.isnull().sum()

btc_market_price                    0
btc_market_cap                      0
btc_n_transactions                  0
btc_n_transactions                  0
btc_cost_per_transaction            0
btc_difficulty                      0
btc_hash_rate                       0
btc_cost_per_transaction_percent    0
dtype: int64

In [57]:
# Splitting the data into training and testing set
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.2,random_state=21)

In [58]:
# Modelling the data
linreg = LinearRegression()  
linreg.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [59]:
# Predicting the test set
y_pred=linreg.predict(X_test)

In [61]:
# Evaluating the model using mean squared error
print("Mean Squared Error: ", metrics.mean_squared_error(y_test, y_pred))

Mean Squared Error:  428.3357597881928
